src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs

```diff
@@ -0,0 +1,102 @@
+		{
+			foreach (var member in typeSymbol.GetDynamicallyAccessedMembers (requiredMemberTypes, declaredOnly)) {
+				switch (member) {
+				case IMethodSymbol method:
+					GetDiagnosticsForMethod (diagnosticContext, method);
+					break;
+				case IFieldSymbol field:
+					GetDiagnosticsForField (diagnosticContext, field);
+					break;
+				case IPropertySymbol property:
+					GetDiagnosticsForProperty (diagnosticContext, property);
+					break;
+				/* Skip Type and InterfaceImplementation marking since doesnt seem relevant for diagnostic generation
+				case ITypeSymbol nestedType:
+					MarkType (diagnosticContext, nestedType);
+					break;
+				case InterfaceImplementation interfaceImplementation:
+					MarkInterfaceImplementation (analysisContext, interfaceImplementation, dependencyKind);
+					break;
+				*/
+				case IEventSymbol @event:
+					GetDiagnosticsForEvent (diagnosticContext, @event);
+					break;
+				}
+			}
+		}
+
+		static void ReportRequiresUnreferencedCodeDiagnostic (in DiagnosticContext diagnosticContext, AttributeData requiresAttributeData, ISymbol member)
+		{
+			var message = RequiresUnreferencedCodeUtils.GetMessageFromAttribute (requiresAttributeData);
+			var url = RequiresAnalyzerBase.GetUrlFromAttribute (requiresAttributeData);
+			diagnosticContext.ReportDiagnostic (DiagnosticId.RequiresUnreferencedCode, member.GetDisplayName (), message, url);
+		}
+
+		static void GetDiagnosticsForMethod (in DiagnosticContext diagnosticContext, IMethodSymbol methodSymbol)
+		{
+			if (methodSymbol.TryGetRequiresUnreferencedCodeAttribute (out var requiresAttributeData))
+				ReportRequiresUnreferencedCodeDiagnostic (diagnosticContext, requiresAttributeData, methodSymbol);
+
+			if (methodSymbol.IsVirtual && methodSymbol.GetDynamicallyAccessedMemberTypesOnReturnType () != DynamicallyAccessedMemberTypes.None)
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			else if (methodSymbol.IsVirtual && methodSymbol.MethodKind is MethodKind.PropertyGet &&
+				(methodSymbol.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None ||
+				methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			} else if (methodSymbol.MethodKind is MethodKind.PropertySet &&
+				  (methodSymbol.GetDynamicallyAccessedMemberTypesOnReturnType () != DynamicallyAccessedMemberTypes.None ||
+				  methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			} else {
```

Comment:
I'm still not sure about this one.

```C#
class TestType
{
    [DAM(Methods)]
    public Type MyType { get; }

    public Type MyTypeWithGetter { [return: DAM(Methods)] get { return null; } }
}

// This should produce 2 warnings
// but the code above will very likely only produce 1 - the first one, not for `MyTypeWithGetter`
typeof (TestType).RequiresAll ();
```

The reason is that on a getter we check if the method itself has DAM, but on a setter we check if the return value has DAM. Return value on setter is void, so that is super weird as well.

Reply:
Added test for these cases